spec/services/activity_service_spec.rb

```diff
@@ -0,0 +1,744 @@
+
+      Timecop.freeze(aug_3) do
+        expect(activity_service.follow_ups).to eq(june_1 => 3, july_1 => 1)
+      end
+    end
+
+    it "can return only the last N periods" do
+      [
+        june_1, june_2, june_3,
+        july_1,
+        aug_1, aug_2
+      ].each do |date|
+        patient = create(:patient, :hypertension, recorded_at: long_ago)
+        create(:blood_pressure, patient: patient, facility: facility_1, recorded_at: date)
+      end
+
+      activity_service = ActivityService.new(facility_1, last: 2)
+
+      Timecop.freeze(aug_3) do
+        expect(activity_service.follow_ups).to eq(july_1 => 1, aug_1 => 2)
+      end
+    end
+
+    it "can return diabetes patients only" do
+      # Excluded patients
+      [
+        june_1, june_2, june_3,
+        july_1,
+        aug_1, aug_2
+      ].each do |date|
+        htn_patient = create(:patient, :hypertension, recorded_at: long_ago)
+        create(:blood_pressure, patient: htn_patient, facility: facility_1, recorded_at: date)
+        create(:blood_sugar, patient: htn_patient, facility: facility_1, recorded_at: date)
+
+        dm_patient_with_bp = create(:patient, :hypertension, recorded_at: long_ago)
+        create(:blood_pressure, patient: dm_patient_with_bp, facility: facility_1, recorded_at: date)
+      end
+
+      # Included patients
+      [
+        june_1,
+        july_1, july_3,
+        aug_1, aug_2
+      ].each do |date|
+        patient = create(:patient, :diabetes, recorded_at: long_ago)
+        create(:blood_sugar, patient: patient, facility: facility_1, recorded_at: date)
+      end
+
+      activity_service = ActivityService.new(facility_1, diagnosis: :diabetes)
+
+      Timecop.freeze(aug_3) do
+        expect(activity_service.follow_ups).to eq(june_1 => 1, july_1 => 2, aug_1 => 2)
+      end
+    end
+
+    it "can return all patients" do
+      [
+        june_1, june_2, june_3,
+        july_1,
+        aug_1, aug_2
+      ].each do |date|
+        patient = create(:patient, :hypertension, recorded_at: long_ago)
+        # We need encounters for these BPs since the generic follow up scope traverses Encounter records
+        create(:blood_pressure, :with_encounter, patient: patient, facility: facility_1, recorded_at: date)
+      end
+
+      [
+        june_1,
+        july_1, july_3,
+        aug_1, aug_2
+      ].each do |date|
+        patient = create(:patient, :diabetes, recorded_at: long_ago)
+        # We need encounters for these blood sugars since the generic follow up scope traverses Encounter records
+        create(:blood_sugar, :with_encounter, patient: patient, facility: facility_1, recorded_at: date)
+      end
+
+      activity_service = ActivityService.new(facility_1, diagnosis: :all)
+
+      Timecop.freeze(aug_3) do
+        expect(activity_service.follow_ups).to eq(june_1 => 4, july_1 => 3, aug_1 => 4)
+      end
+    end
+
+    it "counts multiple visits in a month as one monthly followup" do
```

Comment:
This definition makes sense to me, and seems a lot simpler to correctly implement than the alternatives. I know this is different from some of the other definitions that have been thrown around, but getting this consistent first is great.